mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,328 @@
+- $s_0$ is a reference smoothness value
+- $\kappa$ controls the width of the transition between 0 to 1
+
+Smoothness Indicator Evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: smoothness_indicator
+
+AV RHS Evaluation
+^^^^^^^^^^^^^^^^^
+
+.. autofunction:: av_operator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from pytools.obj_array import obj_array_vectorize
+from meshmode.dof_array import thaw, DOFArray
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from grudge.dof_desc import DD_VOLUME_MODAL, DD_VOLUME
+
+
+def _facial_flux_q(discr, q_tpair):
+    """Compute facial flux for each scalar component of Q."""
+    flux_dis = q_tpair.avg
+    if isinstance(flux_dis, np.ndarray):
+        actx = flux_dis[0].array_context
+        flux_dis = flux_dis.reshape(-1, 1)
+    else:
+        actx = flux_dis.array_context
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    # This uses a central scalar flux along nhat:
+    # flux = 1/2 * (Q- + Q+) * nhat
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_r(discr, r_tpair):
+    """Compute facial flux for vector component of grad(Q)."""
+    flux_dis = r_tpair.avg
+    if isinstance(flux_dis[0], np.ndarray):
+        actx = flux_dis[0][0].array_context
+    else:
+        actx = flux_dis[0].array_context
+
+    normal = thaw(actx, discr.normal(r_tpair.dd))
+
+    # This uses a central vector flux along nhat:
+    # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+    flux_out = flux_dis @ normal
+
+    return discr.project(r_tpair.dd, "all_faces", flux_out)
+
+
+def av_operator(discr, t, eos, boundaries, q, alpha, **kwargs):
```

Comment:
Can we document what `**kwargs` is used for?

Reply:
Will do